API/Controllers/TournamentsController.cs

```diff
@@ -1,34 +1,223 @@
+        }
+
+        // Ensure request is only attempting to perform a replace operation.
+        if (!patch.IsReplaceOnly())
+        {
+            return BadRequest();
+        }
+
+        // Patch and validate
+        patch.ApplyTo(tournament, ModelState);
+        if (!TryValidateModel(tournament))
+        {
+            return BadRequest(ModelState);
+        }
+
+        // Apply patched values to entity
+        TournamentDTO? updatedTournament = await _tournamentsService.UpdateAsync(id, tournament);
+        return Ok(updatedTournament!);
+    }
+
+    /// <summary>
+    /// Submit matches to an existing tournament
+    /// </summary>
+    /// <param name="id">Tournament id</param>
+    /// <param name="wrapper">Match submission data</param>
+    /// <param name="verify">Optionally verify all included matches</param>
+    /// <response code="401">If verify is true and the User does not have match verification privileges</response>
+    /// <response code="404">If a tournament matching the given id does not exist</response>
+    /// <response code="201">Returns location information of the created matches</response>
+    [HttpPost("{id:int}/matches")]
+    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+    [ProducesResponseType(StatusCodes.Status404NotFound)]
+    [ProducesResponseType<IEnumerable<MatchCreatedResultDTO>>(StatusCodes.Status201Created)]
+    public async Task<IActionResult> CreateMatchesAsync(
+        int id,
+        [FromBody] MatchesWebSubmissionDTO wrapper,
+        [FromQuery] bool verify = false
+    )
+    {
+        MatchVerificationSource? verificationSource = null;
+        if (verify)
+        {
+            if (!User.CanVerifyMatches())
+            {
+                return Unauthorized();
+            }
+
+            verificationSource = User.VerificationSource();
+        }
+
+        IEnumerable<MatchCreatedResultDTO>? result =
+            await _matchesService.CreateAsync(
+                id,
+                wrapper.SubmitterId,
+                wrapper.Ids, verify,
+                (int?)verificationSource
+            );
+
+        if (result is null)
+        {
+            return NotFound();
+        }
+
+        // Use no location header for multiple creation
+        return Created((string?)null, result);
+    }
+
+    /// <summary>
+    /// List all matches from a tournament
+    /// </summary>
+    /// <param name="id">Tournament id</param>
+    /// <response code="404">If a tournament matching the given id does not exist</response>
+    [HttpGet("{id:int}/matches")]
+    [ProducesResponseType(StatusCodes.Status404NotFound)]
+    [ProducesResponseType<IEnumerable<MatchDTO>>(StatusCodes.Status200OK)]
+    public async Task<IActionResult> ListMatchesAsync(int id)
     {
-        TournamentDTO res = await _service.CreateOrUpdateAsync(tournament, true);
+        TournamentDTO? result = await _tournamentsService.GetAsync(id);
+        if (result is null)
```

Comment:
I notice we are inconsistently using `is null` and `== null`. I've been using `== null`, just kind of a habit. Do you think `is null` is better going forward, or is there some best practice out there regarding this?

Reply:
> The equality operator can be overridden. "is" keyword is a reference check. "==" will almost always give you what you want. "is" will always give you what you want.

I know we aren't possible overriding the equality operator anywhere, but using `is null` over `==` is generally considered safer. Plus as a python dev it's kind of a force of habit.

Reply:
Sure, I will try to adopt this then. #230